### PR TITLE
Provide configurable clock interval to cater for ledger differences with deduplication tests

### PR DESCRIPTION
**Pull request by dasormeter**
The `CDSimpleDeduplication` and `CDSimpleDeduplicationCommandClient` tests frequently and periodically fail for the distributed ledger Besu and Sawtooth implementations as there is not a sufficient delay between the dedupe period specified in the test (5s) and the wait time given for this to be fully processed before the fresh subsequent command submission comes through (6s)

To rehash what the tests are doing, they are:

1) Submit a command1 with a deduplication time of 5s - expect this to be accepted
2) Submit a dupe of command1 - expect this to be treated as a dupe with corresponding ALREADY_EXISTS grpc response

Wait 6s

3) Submit a command1 again with a deduplication time of 5s -- since the original deduplication time has now expired - expect this to be accepted
4) Submit a dupe of command1 - expect this to be treated as a dupe of step 3

In several of the distributed ledger implementations steps 3 and 4 are not working, because the deduplication window of 1/2 has not expired by the time these are processed.

In this PR the wait time between 2 and 3 has been extended to 2x the deduplication time - so 10s wait for 5s deduplication

As different distributed ledgers will tick time at different configurable intervals, we should also make this ledger clock tick configurable.  This PR also includes the addition of an optional variable to specify ledger clock interval.

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala [hunk 1]
```diff
@@ -19,15 +19,17 @@ import scala.concurrent.Future
 import scala.concurrent.duration.DurationInt
 import scala.util.{Failure, Success}
 
-final class CommandDeduplicationIT extends LedgerTestSuite {
+class CommandDeduplicationIT(ledgerTimeIntervalMs: Int) extends LedgerTestSuite {
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:22]: Why did you remove the `final` qualifier? In which way would you like to extend this class?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala [hunk 2]
```diff
@@ -19,15 +19,17 @@ import scala.concurrent.Future
...
+  private val timeInterval = ledgerTimeIntervalMs / 1000
+  private val deduplicationSeconds = 1 * timeInterval
+  private val deduplicationWindowWait = 2 * timeInterval
 
   test(
-    "CDSimpleDeduplication",
+    "CDSimpleDeduplicationBasic",
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:28]: I thought you already renamed these tests.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala [hunk 3]
```diff
@@ -19,15 +19,17 @@ import scala.concurrent.Future
 import scala.concurrent.duration.DurationInt
 import scala.util.{Failure, Success}
 
-final class CommandDeduplicationIT extends LedgerTestSuite {
+class CommandDeduplicationIT(ledgerTimeIntervalMs: Int) extends LedgerTestSuite {
+  private val timeInterval = ledgerTimeIntervalMs / 1000
+  private val deduplicationSeconds = 1 * timeInterval
+  private val deduplicationWindowWait = 2 * timeInterval
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:25]:
```suggestion
  private val deduplicationSeconds = ledgerTimeIntervalMs / 1000
  private val deduplicationWindowWait = 2 * deduplicationSeconds
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala [hunk 4]
```diff
@@ -104,7 +104,7 @@ object LedgerApiTestTool {
     }
 
     val allTestCaseNames: Set[String] =
-      (Tests.all ++ Tests.retired).flatMap(_.tests).map(_.name).toSet
+      (Tests.all(config) ++ Tests.retired).flatMap(_.tests).map(_.name).toSet
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:107]: You're recreating over and over again all the tests suites across the entire class.

Please create a variable, call `Test.all(config)` once and use that value.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 1 above)
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:22]: Please use `Duration`.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 5]
```diff
@@ -197,6 +197,11 @@ object Cli {
       .action((_, _) => { println(BuildInfo.Version); sys.exit(0) })
       .text("Prints the version on stdout and exit.")
 
+    opt[Int]("ledger-clock-tick-interval")
+      .optional()
+      .action((interval, c) => c.copy(ledgerClockTickIntervalMs = interval))
+      .text("Specify the interval in ms at which the clock of the ledger under test ticks.")
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:203]: It's very difficult to understand the meaning of this flag from it's description and name and why one should choose change it. The name could probably be well suited for the context you have on the ledger you are testing. Please have a second look at the name and description of this option.

@rautenrieth-da Can you have a look at this one? You can probably suggest a more telling name for this.

**Review by stefanobaghino-da — CHANGES_REQUESTED** [review]
Please consider documenting this option. The documentation is located under `//docs/source/tools/ledger-api-test-tool/index.rst`.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 5 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:203]: Perhaps a name like "ledger-clock-granularity"?

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:203]: Reading the code it looks like a timeout, rather than some form of ticks or granularity.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 2 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:28]: will rebase

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 1 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:22]: initially I did this because I was going to do a companion object but not needed now : will revert

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 5 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:203]: it is basically an indicator of ledger clock granularity - the max time you will see between clock ticks on the ledger under test

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala]: I'm not sure I understand. A `final` class can have a companion object.

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala]: Still unadressed.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala (hunk 4 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala]: addressed comment

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 3 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala]: addressed

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 5 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala]: renamed and added better description

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 1 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala]: addressed in latest push now https://github.com/digital-asset/daml/pull/7294/commits/12064ba6e1ff7ca961da400072bf396bbf1b17da

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 6]
```diff
@@ -197,6 +197,11 @@ object Cli {
       .action((_, _) => { println(BuildInfo.Version); sys.exit(0) })
       .text("Prints the version on stdout and exit.")
 
+    opt[Int]("ledger-clock-granularity")
+      .optional()
+      .action((interval, c) => c.copy(ledgerClockGranularityMs = interval))
+      .text("Specify the largest interval in ms that you will see between clock ticks on the ledger under test")
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:203]:
```suggestion
      .text("Specify the largest interval (in ms) that you will see between clock ticks on the ledger under test. The default is 10000ms.")
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala [hunk 7]
```diff
@@ -26,6 +26,7 @@ final case class Config(
     listTestSuites: Boolean,
     shuffleParticipants: Boolean,
     partyAllocation: PartyAllocationConfiguration,
+    ledgerClockGranularityMs: Int
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala:29]: Trailing commas made your diff nicer. Please try to pay it forward. 🙂 

```suggestion
    ledgerClockGranularityMs: Int,
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala [hunk 8]
```diff
@@ -46,5 +47,6 @@ object Config {
     listTestSuites = false,
     shuffleParticipants = false,
     partyAllocation = PartyAllocationConfiguration.ClosedWorldWaitingForAllParticipants,
+    ledgerClockGranularityMs = 10000
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala:50]:
```suggestion
    ledgerClockGranularityMs = 10000,
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala [hunk 9]
```diff
@@ -12,23 +12,22 @@ import com.daml.ledger.test.model.DA.Types.Tuple2
...
 import com.daml.timer.Delayed
-import com.google.protobuf.duration.Duration
 import io.grpc.Status
 
 import scala.concurrent.Future
-import scala.concurrent.duration.DurationInt
 import scala.util.{Failure, Success}
 
-final class CommandDeduplicationIT extends LedgerTestSuite {
+final class CommandDeduplicationIT(ledgerTimeIntervalMs: scala.concurrent.duration.Duration) extends LedgerTestSuite {
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:20]: The duration encapsulates the time unit.

```suggestion
final class CommandDeduplicationIT(ledgerTimeInterval: scala.concurrent.duration.Duration) extends LedgerTestSuite {
```

**Review by SamirTalwar-DA — APPROVED** [review]
Looks sensible to me. Got some suggestions for you though.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 9 above)
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala:20]: Perhaps consider importing `scala.concurrent.duration.Duration` as `ScalaDuration` and `com.google.protobuf.duration.Duration` as `ProtobufDuration` to make things easier to read.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 6 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala]: addressed

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala (hunk 7 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala]: addressed

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala (hunk 9 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala]: addressed

Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandDeduplicationIT.scala]: addressed

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala (hunk 8 above)
Comment (dasormeter) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Config.scala]: addressed

**Comment by dasormeter** [comment]
> Please consider documenting this option. The documentation is located under `//docs/source/tools/ledger-api-test-tool/index.rst`.

@stefanobaghino-da believe all your points have been addressed - can you plase approve / remove requested changes if happy?  thx for the review and improvements

**Review by stefanobaghino-da — APPROVED** [review]
